processor/samplingprocessor/tailsamplingprocessor/processor.go

```diff
@@ -244,37 +241,49 @@ func (tsp *tailSamplingSpanProcessor) makeDecision(id pdata.TraceID, trace *samp
 }
 
 // ConsumeTraceData is required by the SpanProcessor interface.
-func (tsp *tailSamplingSpanProcessor) ConsumeTraces(_ context.Context, td pdata.Traces) error {
+func (tsp *tailSamplingSpanProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
 	tsp.start.Do(func() {
 		tsp.logger.Info("First trace data arrived, starting tail_sampling timers")
 		tsp.policyTicker.Start(1 * time.Second)
 	})
 
-	octds := internaldata.TraceDataToOC(td)
-	var errs []error
-	for _, octd := range octds {
-		if err := tsp.processTraces(octd); err != nil {
-			errs = append(errs, err)
+	resourceSpans := td.ResourceSpans()
+	sampledTraceData := pdata.NewTraces()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		resourceSpan := resourceSpans.At(i)
+		if resourceSpan.IsNil() {
+			continue
 		}
+		tsp.processTraces(ctx, resourceSpan, sampledTraceData)
 	}
-	return componenterror.CombineErrors(errs)
+	return nil
 }
 
-func (tsp *tailSamplingSpanProcessor) processTraces(td consumerdata.TraceData) error {
-
-	// Groupd spans per their traceId to minimize contention on idToTrace
-	idToSpans := make(map[traceKey][]*tracepb.Span)
-	for _, span := range td.Spans {
-		if len(span.TraceId) != 16 {
-			tsp.logger.Warn("Span without valid TraceId", zap.String("SourceFormat", td.SourceFormat))
+func (tsp *tailSamplingSpanProcessor) groupSpansByTraceKey(resourceSpans pdata.ResourceSpans, td pdata.Traces) map[traceKey][]*pdata.Span {
+	idToSpans := make(map[traceKey][]*pdata.Span)
+	ilss := resourceSpans.InstrumentationLibrarySpans()
+	for j := 0; j < ilss.Len(); j++ {
+		ils := ilss.At(j)
+		if ils.IsNil() {
 			continue
 		}
-		tk := traceKey(span.TraceId)
-		idToSpans[tk] = append(idToSpans[tk], span)
+		for k := 0; k < ils.Spans().Len(); k++ {
+			span := ils.Spans().At(k)
+			tk := traceKey(span.TraceID().Bytes())
```

Comment:
`HexBytes()` might be more suitable here.

Reply:
Yes, sorry, `HexString`. Isn't `traceKey` a string-based type? I thought `HexString` would be safer to use as a string, but I guess it's not a big problem here (unless we log it somewhere).